api/types/usertasks/object.go

```diff
@@ -19,77 +19,191 @@
 package usertasks
 
 import (
+	"encoding/binary"
+	"slices"
+	"time"
+
+	"github.com/google/uuid"
 	"github.com/gravitational/trace"
+	"google.golang.org/protobuf/types/known/timestamppb"
 
 	headerv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
 	usertasksv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/usertasks/v1"
 	"github.com/gravitational/teleport/api/types"
 )
 
-// NewUserTask creates a new UserTask object.
-// It validates the object before returning it.
-func NewUserTask(name string, spec *usertasksv1.UserTaskSpec) (*usertasksv1.UserTask, error) {
-	cj := &usertasksv1.UserTask{
+// UserTaskOption defines a function that mutates a User Task.
+type UserTaskOption func(ut *usertasksv1.UserTask)
+
+// WithExpiration sets the expiration of the UserTask resource.
+func WithExpiration(t time.Time) func(ut *usertasksv1.UserTask) {
+	return func(ut *usertasksv1.UserTask) {
+		ut.Metadata.Expires = timestamppb.New(t)
+	}
+}
+
+// NewDiscoverEC2UserTask creates a new DiscoverEC2 User Task Type.
+func NewDiscoverEC2UserTask(spec *usertasksv1.UserTaskSpec, opts ...UserTaskOption) (*usertasksv1.UserTask, error) {
+	taskName := TaskNameForDiscoverEC2(TaskNameForDiscoverEC2Parts{
+		Integration: spec.GetIntegration(),
+		IssueType:   spec.GetIssueType(),
+		AccountID:   spec.GetDiscoverEc2().GetAccountId(),
+		Region:      spec.GetDiscoverEc2().GetRegion(),
+	})
+
+	ut := &usertasksv1.UserTask{
 		Kind:    types.KindUserTask,
 		Version: types.V1,
 		Metadata: &headerv1.Metadata{
-			Name: name,
+			Name: taskName,
 		},
 		Spec: spec,
 	}
+	for _, o := range opts {
+		o(ut)
+	}
 
-	if err := ValidateUserTask(cj); err != nil {
+	if err := ValidateUserTask(ut); err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	return cj, nil
+	return ut, nil
 }
 
+const (
+	// TaskStateOpen identifies an issue with an instance that is not yet resolved.
+	TaskStateOpen = "OPEN"
+	// TaskStateResolved identifies an issue with an instance that is resolved.
+	TaskStateResolved = "RESOLVED"
+)
+
+var validTaskStates = []string{TaskStateOpen, TaskStateResolved}
+
 const (
 	// TaskTypeDiscoverEC2 identifies a User Tasks that is created
 	// when an auto-enrollment of an EC2 instance fails.
 	// UserTasks that have this Task Type must include the DiscoverEC2 field.
 	TaskTypeDiscoverEC2 = "discover-ec2"
 )
 
+// discoverEC2IssueTypes is a list of issue types that can occur when trying to auto enroll EC2 instances.
+var discoverEC2IssueTypes = []string{
```

Comment:
should we move this closer to the types definition so it's easier to keep them aligned?

Reply:
I can change it 👍 
The reasoning here is that I was trying to not add usertasks specific things to `types`.
I was also trying to add generic EC2 Auto Discover failures so that we can use them everywhere, even if not related to usertasks.

Well, thinking about this now, I think it's fine to not pursue one of the above goals.


I'll move the identifiers from `types` into `usertasks`.